### PR TITLE
Fix panic in Glob on empty subject string.

### LINKED ISSUE #1: Possible panic in glob.Glob.
The following input causes `glob.Glob` to panic with `runtime error: slice bounds out of range`:

```Go
package main

import "github.com/ryanuber/go-glob"

func main() {
	glob.Glob("*.gen.go", "")
}
```

```
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0x15a380, 0xc82000a090)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/ryanuber/go-glob.Glob(0x18d2b0, 0x8, 0x0, 0x0, 0xc820000300)
	src/github.com/ryanuber/go-glob/glob.go:61 +0x4cb
main.main()
	src/github.com/shurcooL/play/31/main.go:38 +0x38
exit status 2
```

It happens on [line 54 of glob.go](https://github.com/ryanuber/go-glob/blob/0067a9abd927e50aed5190662702f81231413ae0/glob.go#L54):

```Go
// Trim evaluated text from subj as we loop over the pattern.
idx := strings.Index(subj, part) + len(part) // At this time, subj is "", part is ".gen.go", so idx becomes 6.
subj = subj[idx:] // Since subj is "", doing subj[6:] is a slice bounds out of range panic.
```

I can work on a fix and create a PR to resolve this.

### PR DESCRIPTION
**Pull request by shurcooL**
Fixes #1.

Add tests.

Add test (in form of example) and benchmark for internal `globSequence` utility.

### CODE REVIEW

File: glob.go [hunk 1]
```diff
@@ -57,3 +62,14 @@ func Glob(pattern, subj string) bool {
 	// All parts of the pattern matched
 	return true
 }
+
+// globSequence reports if pattern consits of glob characters only.
+// I.e., it returns true for "", "*", "**", "***", etc., and false for all other patterns.
+func globSequence(pattern string) bool {
+	for i := 0; i < len(pattern); i++ {
+		if pattern[i] != '*' {
```
Comment (renfredxh) [inline glob.go]: Should this be `pattern[i] != GLOB`? See how `const GLOB` is used in other places in this file.

Comment (shurcooL) [inline glob.go]: I wanted to do that, but couldn't.

`GLOB` is a `string` type. I need a `byte` here.

Comment (shurcooL) [inline glob.go]: I could do `GLOB[0]`, or declare `const globByte = '*'` instead, but I thought simply doing '*' would be okay.

Comment (renfredxh) [inline glob.go]: Got it, I saw your note about the performance implications of this as well. LGTM

**Comment by ryanuber** [comment]
Welp that is pretty terrible! Thanks for finding this and fixing it.

I think we can actually do an even more simple fix though. On line 43..45, let's just remove the condition. If you are on the last chunk, you either have to match a trailing glob, or the suffix of the pattern. We should also switch the `return true` at the very end to `return false` and note that we should never reach it.

If we use the above approach we are actually just removing code which should speed us up a little bit, although we are talking about nanoseconds here. Let's definitely bring in your behavior tests with it.

Thanks again @shurcooL! Great work.

**Comment by shurcooL** [comment]
Thanks for review, I will try that approach and update this PR.

**Comment by shurcooL** [comment]
> We should also switch the `return true` at the very end to `return false` and note that we should never reach it.

If you're confident that it's not possible to ever reach it, I'd argue it's a better practice to do:

```Go
panic("unreachable")
```

With the implicit understanding that it should be impossible to get there, and you ever do, it's a bug or regression in the library that should be fixed.

E.g., something like [this](https://github.com/shurcooL/htmlg/blob/3da295397738f395c2ca5fd5570f01a9fcea3be3/htmlg.go#L46-L49), [this](https://github.com/shurcooL/vfsgen/blob/6e5fb4d6aad9620b57c54f045c7d738a64ff7563/test/test_vfsdata_test.go#L105-L106) or [this](https://github.com/shurcooL/vfsgen/blob/6e5fb4d6aad9620b57c54f045c7d738a64ff7563/test/test_vfsdata_test.go#L122-L123).

The motivation is that if you're 100% sure the logic should never get there in any situation, and it happens anyway, it's better to panic than to silently return incorrect results. It's also more readable and clear to other people that certain code is not a part of normal execution flow (otherwise they might think "how would the code ever get here? am I missing something?").

That said, it might be possible and even better to simply rearrange the logic so that a `panic("unreachable")` is not needed. For example, arrange so `case end` simply breaks out of the for loop, and have a final return statement that is executed. I will leave this to you.

**Comment by shurcooL** [comment]
@ryanuber, I've implemented the change you suggested in a4a902f3562996a1cecd98ed1913f375fbba150c, PTAL.

**Comment by ryanuber** [comment]
@shurcooL great suggestion for the panic, I agree. This looks great. Let's pull it in.

**Comment by shurcooL** [comment]
All right, I've squashed into 1 commit and just added [2 more](https://github.com/ryanuber/go-glob/blob/87ce0b27039818fd8036c1d26e4d84e90e9c3648/glob_test.go#L34-L35) test cases to `TestEmptySubject`.

**Comment by ryanuber** [comment]
:ship:

**Comment by ryanuber** [comment]
Awesome, thanks again @shurcooL! I appreciate it.

**Pull request merged**